### ISSUE TITLE
Connection timeout pulling big containers

### ISSUE THREAD

**Issue by victorsndvg** [singularityhub/sregistry#82]
Hi @vsoch,

Sorry again for bringing this nginx errors ... but I'm not very experienced with this web services and I think this could be easy for you.  :P

I'm trying to pull a 2.3Gb image and I always get the following error:
```
nginx_1   | 2018/01/18 16:46:59 [error] 6#6: *3 upstream timed out (110: Connection timed out) while reading response header from upstream, client: XXXXXX, server: localhost, request: "GET /containers/8/download/3c548175-fd5c-4033-b667-be836f0c7f4b HTTP/1.1", upstream: "uwsgi://172.17.0.4:3031", host: "XXXXXXX2
```
The connections are  always closed after 60s.

This is the current nginx.conf file

```
server {
  listen                *:80;
  server_name           localhost;

  client_max_body_size 8000M;
  client_body_buffer_size 2000M;
  client_body_timeout 900;
  send_timeout 900;

  add_header X-Clacks-Overhead "GNU Terry Pratchett";
  add_header X-Clacks-Overhead "GNU Terry Pratchet";
  add_header Access-Control-Allow-Origin *;
  add_header 'Access-Control-Allow-Credentials' 'true';
  add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
  add_header 'Access-Control-Allow-Headers' 'Authorization,DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type';

  location /images {
    alias /var/www/images;
  }

  location / {
    include /etc/nginx/uwsgi_params.par;
    uwsgi_pass uwsgi:3031;
  }

  location /static {
    alias /var/www/static;
  }
}

server {

        listen 443;
        server_name localhost;

        root html;
        client_max_body_size 8000M;
        client_body_buffer_size 2000M;
        client_body_timeout 900;
        send_timeout 900;

        add_header X-Clacks-Overhead "GNU Terry Pratchett";
        add_header X-Clacks-Overhead "GNU Terry Pratchet";
        add_header Access-Control-Allow-Origin *;
        add_header 'Access-Control-Allow-Credentials' 'true';
        add_header 'Access-Control-Allow-Methods' 'GET, POST, OPTIONS';
        add_header 'Access-Control-Allow-Headers' 'Authorization,DNT,X-CustomHeader,Keep-Alive,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type';


        ssl on;
        ssl_certificate XXXXXXXXXXXXX.crt;
        ssl_certificate_key XXXXXXXXXXX.pem;
        ssl_session_timeout 5m;
        ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
        ssl_ciphers ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA;
        ssl_session_cache shared:SSL:50m;
        ssl_dhparam /etc/ssl/certs/dhparam.pem;
        ssl_prefer_server_ciphers on;

        location /images {
            alias /var/www/images;
        }

        location /static {
            alias /var/www/static;
        }

        location / {
            include /etc/nginx/uwsgi_params.par;
            uwsgi_pass uwsgi:3031;
        }

}
```

Thanks in advance!

**Comment by vsoch** [comment]
Try increasing the timeout.

**Comment by victorsndvg** [comment]
After increasing them to 9000

```
        client_body_timeout 9000;
        send_timeout 9000;
        proxy_set_header Connection "";
        uwsgi_read_timeout 9000;
        uwsgi_send_timeout 9000;
        uwsgi_request_buffering on;
```

I get the following error:

```
uwsgi_1   | [pid: 50|app: 0|req: 1/1] 10.118.68.13 () {38 vars in 567 bytes} [Thu Jan 18 11:22:58 2018] GET /api/container/mso4sc/salome:8.3.0 => generated 1065 bytes in 778 msecs (HTTP/1.1 200) 5 headers in 159 bytes (1 switches on core 0)
nginx_1   | 10.118.68.13 - - [18/Jan/2018:17:22:58 +0000] "GET /api/container/mso4sc/salome:8.3.0 HTTP/1.1" 200 1065 "-" "Python-urllib/2.6" "-"
uwsgi_1   | [pid: 50|app: 0|req: 2/2] 10.118.68.13 () {38 vars in 617 bytes} [Thu Jan 18 11:22:58 2018] GET /containers/8/download/3c548175-fd5c-4033-b667-be836f0c7f4b => generated 850 bytes in 133292 msecs (HTTP/1.1 500) 4 headers in 142 bytes (1 switches on core 0)
nginx_1   | 10.118.68.13 - - [18/Jan/2018:17:25:12 +0000] "GET /containers/8/download/3c548175-fd5c-4033-b667-be836f0c7f4b HTTP/1.1" 500 850 "-" "Python-urllib/2.6" "-"
```

Note that in both cases it fails after 1 min or 2, that is less than the specified timeouts. Are there other timeouts involved that I'm not taking into account? If not, any other idea?

**Comment by vsoch** [comment]
hey @victorsndvg apologies for delay - another day when I'm away from internet. The error appears to be 500, which is a server error. How did you come about setting those other uwsgi settings? I would guess there is something that is conflicting, but I myself don't know much beyond the basics to have inutition about it. What I would try doing is:

 1. Add debugging code [here](https://github.com/singularityhub/sregistry/blob/master/shub/apps/main/views/download.py#L110) so you get a print of everything going on, and better, pickle stuff and save them.
 2. Restart the server, trigger the error to print the debugging stuff and then create the pickle 
 3. shell into the container
 4. python manage.py shell
 5. load the pickle to see if you can re-create the 500 errorop
 
Hopefully that is a start! If this only happens with large containers that is a hint that it's (still) some kind of timeout or mismatched setting. Apologies I don't know more "hiding places" for these settings, beyond the nginx.conf and uwsgi.ini.

**Comment by victorsndvg** [comment]
It's being hard ...
Next thing I see is when the pull is successful, I see the following message:
```
nginx_1   | 2018/01/19 12:59:56 [warn] 5#5: *3 an upstream response is buffered to a temporary file /var/cache/nginx/uwsgi_temp/1/00/0000000001 while reading upstream, client: 10.118.66.1, server: localhost, request: "GET /containers/9/download/5b427198-beb7-44e7-a0ca-74400b444e73 HTTP/1.1", upstream: "uwsgi://172.17.0.4:3031", host: "XXXXXXXXXXXX"
```
When it fails, this is not shown. What happens in the middle?

Maybe some buffers configuration?

**Comment by vsoch** [comment]
Oh that might be! Do you still have the `client_body_buffer_size` at 2000?

**Comment by victorsndvg** [comment]
Nope, since yesterday is `client_body_buffer_size at 7000` (is a 8Gb RAM VM).

**Comment by vsoch** [comment]
Why are the containers so large? Even with squasfs?

**Comment by victorsndvg** [comment]
Not all of them, but there are big stand alone apps or some with a bunch of dependencies...

**Comment by dctrud** [comment]
Some of the biocontainers built into singularity images at https://depot.galaxyproject.org/singularity/ are 2GB or more. Also, if you were to build a single container holding all tools and tool-specific reference data for a Nextflow workflow that can be pretty big (more than 2GB).

**Comment by vsoch** [comment]
hey @victorsndvg ! Can you try adding a socket timeout to the uwsgi.ini? A line like:

```
socket-timeout = 60
```

**Comment by victorsndvg** [comment]
Hi @vsoch , I'm on the road again!

I already tried this option without success... The issue still here. No idea about how to deal with it ...

@dctrud are you also experiencing this issue?

**Comment by dctrud** [comment]
Hi @victorsndvg - am afraid I haven't experienced the issue, but I am not using the same front end proxy setup as is in the sregistry dockerfile. For consistency we use gunicorn and apache across services.

**Comment by vsoch** [comment]
@dctrud are you able to peek at the metadata from a request that might be used to help compare with @victorsndvg configuration?

**Comment by dctrud** [comment]
Hi all - I don't think there's anything that's useful to compare from my logs - it just works. 200 OK codes all round, so I did a bit more digging...

@victorsndvg - Looking back through the above, the logs you have show that uwsgi generated very little output in terms of bytes during the relatively long time the request ran. We can probably put in massive timeouts to make it work, but that's not optimal, and you are likely to hit problems when download size goes close to, or larger than RAM....

I took a look into the sregistry download code and noted that the container download is being sent out via plain HttpResponse. This results in the file being read into RAM in its entirety before anything gets sent out - and from your logs it's timing out during this process, before any data is being sent. There are 2 better ways to handle large files in Django:

1. Use StreamingHttpResponse and the basehttp FileWrapper - this will mean the uwsgi app doesn't read the entire file into RAM - it streams it off disk straight to nginx, then out down the http connection.

2. Use xsendfile (django-sendfile has a good implementation supporting nginx) which offloads the serving of the download to Nginx - which will stream it out directly.

I've done both of these in systems at work that need to serve files that could be in the 100s of GBs without issue. 

@vsoch I can try and to a PR to address this on the weekend.

**Comment by victorsndvg** [comment]
Yeah, @dctrud , I've been looking to the RAM consumption and it increases a lot while preparing to send a container. I've been reading about the solutions you are proposing and it seems to be the right way to deliver big files, but I don't know how to implement them.

It will be great if you can implement them!! Thanks! 👍 

If not I can try to implement the @vsoch  deployment

**Comment by dctrud** [comment]
@victorsndvg - if you can try the PR #90 changes I think they are likely to fix your issues.... and if they don't, the logging change in the PR will give additional useful info via `docker-compose logs uwsgi` for further troubleshooting.

**Comment by vsoch** [comment]
I'm testing too! Holy crap that was fast!
